cfpb/consumerfinance.gov
# Filterable list refactor and optional filterable list controls

This PR refactors the implementation of filterable lists on the website by moving some of their configuration off of the FilterableList StreamField block and on to the page model instances directly, as new fields on the FilterableListMixin model class.

After this change, page types that inherit from FilterableListMixin now have a new "Filtering" panel in the Wagtail editor that can be used to configure how the search is done and how results appear on the page. Page types like these, including BrowseFilterablePage and SublandingFilterablePage and their descendants, now always include search results after their content fields but before the sidefoot field.

Wagtail editors can now choose whether or not to add the FilterableList block to the page; if they do not, the filter controls block will not appear at all, even though the results still will. This addresses internal D&CP#139.

This PR does not touch the logic of executing filterable list searches (the search object and implementation below, including OpenSearch). This refactor should make it easier to do that in future should we want to do so.

This PR includes the migration-related commits from #7812.

## How to test this PR

I've verified the following:

- Python tests pass both with `tox` and skipping Django migrations using `SKIP_DJANGO_MIGRATIONS=1 tox`.
- Functional tests pass against both production and test databases using `yarn cypress run`.

Another thing to test is removing the filter from a page and confirming that results and pagination still work as expected. Edit, for example, [the blog](http://localhost:8000/admin/pages/1491/edit/) and delete the FL block from the content field. Publish the page and confirm that you can still paginate through the results. Note one expected weirdness: if there are links to, say, topic tags, as there are on the blogs, these query strings will still work, and you'll still land on a proper results page, there just won't be any filter visible. This case should ideally never happen the way we set up our filterable list pages but I'm open to suggestions here @csebianlander.

## Screenshots

The new "Filtering" tab:

<img width="500" alt="image" src="https://github.com/cfpb/consumerfinance.gov/assets/654645/1b39ef74-11fc-4bfa-b47b-e0fccfffb00c">

Simplified and reorganized Filterable List configuration:

<img width="400" alt="image" src="https://github.com/cfpb/consumerfinance.gov/assets/654645/8f35e922-b805-4d90-ac40-93bb230fb01c">
<img width="400" alt="image" src="https://github.com/cfpb/consumerfinance.gov/assets/654645/4202910f-cff6-4f43-a405-06a3fce1ff32">

## Notes and todos

- @csebianlander: how do you feel about "Filtering" for the heading of the new tab in the editor?
- @csebianlander: Note above that now filterable list results always appear after a page's content field. This means that, if a filterable list controls block gets added to a page, it will need to be last. I believe this is the case for almost all of our pages - we always end with the results - except for [the OAA page](http://localhost:8000/administrative-adjudication-proceedings/administrative-adjudication-docket/).
   
   |Before|After|
   |-|-|
   |<img width="883" alt="image" src="https://github.com/cfpb/consumerfinance.gov/assets/654645/74dda03a-95a6-434f-92f6-94d202610fb2">...<img width="887" alt="image" src="https://github.com/cfpb/consumerfinance.gov/assets/654645/8896e88b-1322-4c9f-842d-a44480219ed9">|<img width="887" alt="image" src="https://github.com/cfpb/consumerfinance.gov/assets/654645/0422e7b7-3b6e-46f8-9ee2-32d964d6f0fa">|
   
   Any concerns here?
- @willbarton: I had to do a bit of Django migration editing in v1.0001 in order to convert FilterableListMixin from a pure Python class to an abstract Django base model class. I feel good about my testing above but perhaps you could also review this to make sure it seems reasonable.
- @anselmbradford @jenn-franklin: Because this change splits the FL controls from the results, it adds slightly more space between them:
   
   |Before|After|
   |-|-|
   |<img width="888" alt="image" src="https://github.com/cfpb/consumerfinance.gov/assets/654645/cf32b1e5-703e-4ed4-ae3e-fe26214c2980">|<img width="883" alt="image" src="https://github.com/cfpb/consumerfinance.gov/assets/654645/42a939f3-1bd1-4a4d-ad76-c79c6591cf4e">|
   |<img width="888" alt="image" src="https://github.com/cfpb/consumerfinance.gov/assets/654645/38ab44f1-4ab6-48cf-b92c-c25fef19bcde">|<img width="885" alt="image" src="https://github.com/cfpb/consumerfinance.gov/assets/654645/0183ccb5-3a31-40d2-8a8a-25c674ea1db6">|
   
   Any strong feelings about this? I implemented the results block as a `<div class="block">` around a `<div class="o-filterable-list-results">`; does this make sense?

## Checklist
